validator_client/src/beacon_node_fallback.rs

```diff
@@ -0,0 +1,424 @@
+                    .await;
+            }
+        }
+    }
+
+    /// Run `func` against each candidate in `self`, returning immediately if a result is found.
+    /// Otherwise, return all the errors encountered along the way.
+    ///
+    /// First this function will try all nodes with a suitable status. If no candidates are suitable
+    /// or all the requests fail, it will try updating the status of all unsuitable nodes and
+    /// re-running `func` again.
+    pub async fn first_success<'a, F, O, Err, R>(
+        &'a self,
+        require_synced: RequireSynced,
+        func: F,
+    ) -> Result<O, AllErrored<Err>>
+    where
+        F: Fn(&'a BeaconNodeHttpClient) -> R,
+        R: Future<Output = Result<O, Err>>,
+    {
+        let mut errors = vec![];
+        let mut to_retry = vec![];
+
+        // First pass: try `func` on all ready candidates.
+        for candidate in &self.candidates {
+            if let Err(e) = candidate.status(require_synced).await {
+                // This client was not ready on the first pass, we might try it again later.
+                to_retry.push(candidate);
+                errors.push((candidate.beacon_node.to_string(), Error::Unavailable(e)));
+            } else {
+                inc_counter_vec(&ENDPOINT_REQUESTS, &[candidate.beacon_node.as_ref()]);
+
+                // There exists a race condition where `func` may be called when the candidate is
+                // actually not ready. We deem this an acceptable inefficiency.
+                match func(&candidate.beacon_node).await {
+                    Ok(val) => return Ok(val),
+                    Err(e) => {
+                        // If we have an error on this function, make the client as not-ready.
+                        //
+                        // There exists a race condition where the candidate may have been marked
+                        // as ready between the `func` call and now. We deem this an acceptable
+                        // inefficiency.
+                        candidate.set_offline().await;
+                        errors.push((candidate.beacon_node.to_string(), Error::RequestFailed(e)));
+                        inc_counter_vec(&ENDPOINT_ERRORS, &[candidate.beacon_node.as_ref()]);
+                    }
+                }
+            }
+        }
+
+        // Second pass: try again, attempting to make non-ready clients become ready.
+        for candidate in to_retry {
+            // If the candidate hasn't luckily transferred into the correct state in the meantime,
+            // force an update of the state.
+            let new_status = match candidate.status(require_synced).await {
+                Ok(()) => Ok(()),
+                Err(_) => {
+                    candidate
+                        .refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
+                        .await
+                }
+            };
+
+            if let Err(e) = new_status {
```

Comment:
If `require_synced == RequireSync::No` shouldn't we ignore here a sync error and proceed with the `else` branch?

Reply:
Nice catch! I've addressed this in f6e148620.